File: src/ai/grid.js
(function(global) {
    class Grid {
        constructor(width, height) {
            this.width = width;
            this.height = height;
            // 0: Empty, 1: Food, 2: My Body, 3: Enemy Body
            this.cells = new Int8Array(width * height).fill(0);
        }

        _idx(x, y) {
            return y * this.width + x;
        }

        get(x, y) {
            if (x < 0 || x >= this.width || y < 0 || y >= this.height) return 9; // Out of bounds
            return this.cells[this._idx(x, y)];
        }

        set(x, y, val) {
            if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
                this.cells[this._idx(x, y)] = val;
            }
        }

        isSafe(x, y) {
            const val = this.get(x, y);
            // Safe: Empty (0) or Food (1)
            return val === 0 || val === 1;
        }

        clone() {
            const newGrid = new Grid(this.width, this.height);
            newGrid.cells.set(this.cells);
            return newGrid;
        }

        static fromState(state) {
            // state expects: { cols, rows, food, pBody, aBody }
            const g = new Grid(state.cols, state.rows);
            
            // Place Food
            if (state.food) {
                state.food.forEach(f => g.set(f.x, f.y, 1));
            }

            // Place Player (Me) - Mark everything as 2
            if (state.pBody) {
                state.pBody.forEach(p => {
                    g.set(p.x, p.y, 2); 
                });
            }

            // Place AI (Enemy) - Mark everything as 3
            if (state.aBody) {
                state.aBody.forEach(p => {
                    g.set(p.x, p.y, 3);
                });
            }

            return g;
        }
    }

    if (typeof module !== 'undefined' && module.exports) {
        module.exports = Grid;
    } else {
        global.SnakeAI = global.SnakeAI || {};
        global.SnakeAI.Grid = Grid;
    }
})(this);

File: rust/crates/snake-ai/src/grid.rs
use snake_domain::Point;

#[derive(Debug, Clone)]
pub struct Grid {
    pub width: i32,
    pub height: i32,
    pub cells: Vec<i8>,
}

impl Grid {
    pub fn new(width: i32, height: i32) -> Self {
        Self {
            width,
            height,
            cells: vec![0; (width * height) as usize],
        }
    }

    #[inline]
    pub fn idx(&self, x: i32, y: i32) -> usize {
        (y * self.width + x) as usize
    }

    #[inline]
    pub fn get(&self, x: i32, y: i32) -> i8 {
        if x < 0 || y < 0 || x >= self.width || y >= self.height {
            return 9;
        }
        self.cells[self.idx(x, y)]
    }

    #[inline]
    pub fn set(&mut self, x: i32, y: i32, val: i8) {
        if x < 0 || y < 0 || x >= self.width || y >= self.height {
            return;
        }
        let idx = self.idx(x, y);
        self.cells[idx] = val;
    }

    #[inline]
    pub fn is_safe(&self, x: i32, y: i32) -> bool {
        matches!(self.get(x, y), 0 | 1)
    }

    pub fn from_state(cols: i32, rows: i32, food: &[Point], my_body: &[Point], enemy_body: &[Point]) -> Self {
        let mut g = Self::new(cols, rows);
        for f in food {
            g.set(f.x, f.y, 1);
        }
        for p in my_body {
            g.set(p.x, p.y, 2);
        }
        for p in enemy_body {
            g.set(p.x, p.y, 3);
        }
        g
    }
}
